### ISSUE TITLE
VideoStreamDecoder.TryDecodeNextFrame (sample) leaks unmanaged memory

### ISSUE THREAD

**Issue by springy76** [Ruslan-B/FFmpeg.AutoGen#112]
I already disabled/tested everything else, only within `TryDecodeNextFrame` unmanaged memory gets leaked. Definitley not the entire frame buffers but when 16 threads are processing frames then I loose 1GB per minute -- managed RAM keeps constant at ~250KB.

**Comment by springy76** [comment]
Seems every uninteresting frame filtered by `while (_pPacket->stream_index != _streamIndex)` gets leaked.

**Comment by Ruslan-B** [comment]
Looking into TryDecodeNextFrame and trying to compare code with examples in C - I see quite ofter they are using av_free_packet after each av_read_frame but av_free_packet  is marked as obsolete with advice to use av_packet_unref instead.

I tried to rewrite TryDecodeNextFrame in order to do so per packet and I see the app consumes a littel bit less memory:

```csharp
  public bool TryDecodeNextFrame(out AVFrame frame)
        {
            ffmpeg.av_frame_unref(_pFrame);
            int error;
            do
            {
                do
                {
                    ffmpeg.av_packet_unref(_pPacket);

                    error = ffmpeg.av_read_frame(_pFormatContext, _pPacket);
                    if (error == ffmpeg.AVERROR_EOF)
                    {
                        frame = *_pFrame;
                        return false;
                    }

                    error.ThrowExceptionIfError();
                } while (_pPacket->stream_index != _streamIndex);

                ffmpeg.avcodec_send_packet(_pCodecContext, _pPacket).ThrowExceptionIfError();

                error = ffmpeg.avcodec_receive_frame(_pCodecContext, _pFrame);
            } while (error == ffmpeg.AVERROR(ffmpeg.EAGAIN));

            error.ThrowExceptionIfError();
            frame = *_pFrame;
            return true;
        }
```
Please let me know if it going to make any difference for you.

**Comment by springy76** [comment]
I meanwhile restructured the code as below to handle (drain) any buffered frames and also not to stop/exit on invalid packets:

```cs
public bool TryDecodeNextFrame(out AVFrame frame)
{
	ffmpeg.av_frame_unref(this._pFrame);
	int error;

	// repeated-try avcodec_receive_frame until enough packets haven been send
	while ( true )
	{
		// try read frame; maybe last avcodec_send_packet resulted in multiple frames to be read; will respond with EAGAIN if more packets are needed
		error = ffmpeg.avcodec_receive_frame(this._pCodecContext, this._pFrame);
		if ( error != ffmpeg.AVERROR(ffmpeg.EAGAIN) )
		{
			frame = *this._pFrame;
			if ( error == ffmpeg.AVERROR_EOF )
				return false;

			error.ThrowExceptionIfError();
			return true;
		}

		try
		{
			// feed all stream-matching packets to decoder
			while ( true )
			{
				error = ffmpeg.av_read_frame(this._pFormatContext, this._pPacket);
				if ( error == ffmpeg.AVERROR_EOF )
				{
					// no more packets to read -> trigger draining/flushing of remaining frames
					ffmpeg.avcodec_send_packet(this._pCodecContext, null).ThrowExceptionIfError();
					break; // don't throw error, just exit packet-feed-loop
				}

				error.ThrowExceptionIfError();

				if ( this._pPacket->stream_index == this._streamIndex )
				{
					var sendPacketResult = ffmpeg.avcodec_send_packet(this._pCodecContext, this._pPacket);
					if ( sendPacketResult == 0 )
						break;

					// no reason to abort/crash (it was just an invalid packet from demuxer) -> just retrieve and feed next packet; most of time the code will resume without problems
					var errorMsg = FFmpegHelper.av_strerror(sendPacketResult);
					Trace.TraceError(errorMsg);
				}

				// unref all packets, no matter if sent to avcodec or ignored
				ffmpeg.av_packet_unref(this._pPacket);
			}
		}
		finally
		{
			ffmpeg.av_packet_unref(this._pPacket);
		}
	}
}
```

**Comment by Ruslan-B** [comment]
Well I'm ported code from the C examples of FFmpeg and didn't invent it myself. On the other hand they usually operating withing a single stack frame without any OOP :). I try to reintegrate this code into example, I hope it is okay with you, besides I might change it as well.

**Comment by springy76** [comment]
Feel free to do whatever fits best with this code. I tested it against some thousand files with great success.

**Comment by Ruslan-B** [comment]
I did memory profiling using recent dotMemory from JetBrains and I failing to see any leaks in the example code in current state. I have a stable plato and about 200MB of total ram consumption. 
I do realize though in different scenarios leaks might appear as I found from official examples there is no single way to feed data into decoder. So I'll keep code in current state as it is already complex enough to comprehend.
